File: solutions/295.find-median-from-data-stream.java
/*
 * @lc app=leetcode id=295 lang=java
 *
 * [295] Find Median from Data Stream
 */

// @lc code=start
class MedianFinder {
    Queue<Integer> lq = new PriorityQueue<>((a,b)->b-a); // max-heap
    Queue<Integer> rq = new PriorityQueue<>((a,b)->a-b); // min-heap

    public MedianFinder() {
        
    }
    
    public void addNum(int num) {
        if(lq.isEmpty()) lq.offer(num);
        else if(num < lq.peek()) lq.offer(num);
        else rq.offer(num);
        
        // make balance btw lq , rq that size <=1
        if(lq.size()+2 == rq.size()) lq.offer(rq.poll());
        else if(lq.size() == rq.size()+2) rq.offer(lq.poll());
    }
    
    public double findMedian() {
        if(lq.size()==rq.size()) return (lq.peek() + rq.peek())/2.0;
        if(lq.size()<rq.size()) return rq.peek();
        return lq.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
// @lc code=end

File: solutions/295.find-median-from-data-stream.rb
class MedianFinder
  attr_reader :lq, :rq

  def initialize
    @lq = PriorityQueue.new { _1 > _2 } # max-top heap
    @rq = PriorityQueue.new { _1 < _2 } # min-top heap
  end

  #   :type num: Integer
  #   :rtype: Void
  def add_num(num)
    if rq.empty? || num < rq.next
      lq.push(num.to_f, num)
    else
      rq.push(num.to_f, num)
    end

    if lq.size + 2 == rq.size
      pop = rq.pop
      lq.push(pop, pop)
    elsif lq.size == rq.size + 2
      pop = lq.pop
      rq.push(pop, pop)
    end
  end

  #   :rtype: Float
  def find_median
    return lq.next if lq.size > rq.size
    return rq.next if lq.size < rq.size

    (lq.next + rq.next) / 2
  end
end

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder.new()
# obj.add_num(num)
# param_2 = obj.find_median()
